### PR TITLE
ATO-945: Create alarms for messages not being read from SPOT request queue

### PR DESCRIPTION
**Pull request by ethanmills**
Thresholds based on previous year of metrics. There was one P1 incident which should have triggered this alarm, then a couple of spikes to around 40 seconds which should have been investigated but not triggered pager duty.

## How to review

1. Ensure metric queries and thresholds look right
2. Review linked runbooks

### CODE REVIEW

**Comment by ethanmills** [comment]
I've run a terraform plan of this but not deployed it

File: ci/terraform/oidc/alerts.tf [hunk 1]
```diff
@@ -30,10 +30,44 @@ resource "aws_cloudwatch_metric_alarm" "spot_request_sqs_dlq_cloudwatch_alarm" {
   dimensions = {
     QueueName = aws_sqs_queue.spot_request_dead_letter_queue.name
   }
-  alarm_description = "${var.dlq_alarm_threshold} or more messages have appeared on the ${aws_sqs_queue.spot_request_dead_letter_queue.name}.ACCOUNT: ${data.aws_iam_account_alias.current.account_alias}"
+  alarm_description = "${var.dlq_alarm_threshold} or more messages have appeared on the ${aws_sqs_queue.spot_request_dead_letter_queue.name}. ACCOUNT: ${data.aws_iam_account_alias.current.account_alias}. Runbook: https://govukverify.atlassian.net/wiki/x/DYDMBgE"
```
Comment (billrdunn) [inline ci/terraform/oidc/alerts.tf:33]: Unfinished sentence in runbook?  "Users affected will be returned to the relying party with..."

File: ci/terraform/oidc/alerts.tf [hunk 2]
```diff
@@ -30,10 +30,44 @@ resource "aws_cloudwatch_metric_alarm" "spot_request_sqs_dlq_cloudwatch_alarm" {
...
+  alarm_actions     = [data.aws_sns_topic.slack_events.arn]
+}
+
+resource "aws_cloudwatch_metric_alarm" "spot_request_sqs_cloudwatch_p1_alarm" {
+  alarm_name          = replace("${var.environment}-spot-request-queue-alarm", ".", "")
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = "ApproximateAgeOfOldestMessage"
+  namespace           = "AWS/SQS"
+  period              = "300"
```
Comment (billrdunn) [inline ci/terraform/oidc/alerts.tf:60]: Is 5 minutes quite a long time to wait for a P1?

File: ci/terraform/oidc/alerts.tf [hunk 3]
```diff
@@ -30,10 +30,44 @@ resource "aws_cloudwatch_metric_alarm" "spot_request_sqs_dlq_cloudwatch_alarm" {
...
+
+  dimensions = {
+    QueueName = aws_sqs_queue.spot_request_queue.name
+  }
+  alarm_description = "Age of the oldest message on ${aws_sqs_queue.spot_request_queue.name} exceeds 10 seconds. ACCOUNT: ${data.aws_iam_account_alias.current.account_alias}. Runbook: https://govukverify.atlassian.net/wiki/x/VIFoCAE"
+  alarm_actions     = [data.aws_sns_topic.slack_events.arn]
+}
+
+resource "aws_cloudwatch_metric_alarm" "spot_request_sqs_cloudwatch_p1_alarm" {
+  alarm_name          = replace("${var.environment}-spot-request-queue-alarm", ".", "")
```
Comment (billrdunn) [inline ci/terraform/oidc/alerts.tf:55]:
```suggestion
  alarm_name          = replace("${var.environment}-P1-spot-request-queue-alarm", ".", "")
```

**Comment by billrdunn** [comment]
Do we not trigger a pager duty alert for the DLQ alarm because this means SPOT will be alarming?

**Comment by billrdunn** [comment]
ATO-945 says "Count the number of SPOT requests which time out and alert on this metric". Is there a reason you went for age of oldest message instead?

**Comment by ethanmills** [comment]
> [ATO-945](https://govukverify.atlassian.net/browse/ATO-945) says "Count the number of SPOT requests which time out and alert on this metric". Is there a reason you went for age of oldest message instead?

I'll do both sooner or later, just starting with this one. This is ATO-946 really, and I'll pick up 947 at some other point

**Review by billrdunn — APPROVED** [review]

**Comment by ethanmills** [comment]
> Do we not trigger a pager duty alert for the DLQ alarm because this means SPOT is not down and will be alarming?

That's a good justification, but to be honest it's just because it wasn't in the scope of the ticket

**Pull request merged**